ipc/api/src/checkpoint.rs

```diff
@@ -86,3 +96,81 @@ pub struct BottomUpCheckpoint {
     /// The list of messages for execution
     pub msgs: Vec<IpcEnvelope>,
 }
+
+fn deserialize_vec_bytes_from_vec_hex<'de, D>(
+    deserializer: D,
+) -> anyhow::Result<Vec<Vec<u8>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let data = <Vec<String>>::deserialize(deserializer)?;
+
+    let mut ret = Vec::new();
+    for s in data {
+        let s =
+            hex::decode(&s).map_err(|_| D::Error::custom(format!("cannot decode hex: {}", s)))?;
+        ret.push(s);
+    }
+    Ok(ret)
+}
+
+pub fn serialize_vec_bytes_to_hex<T: AsRef<[u8]>, S>(data: &[T], s: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let values = data.iter().map(hex::encode).collect::<Vec<_>>();
+
+    let mut seq = s.serialize_seq(Some(values.len()))?;
+    for element in values {
+        seq.serialize_element(&element)?;
+    }
+    seq.end()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::address::IPCAddress;
+    use crate::checkpoint::{
+        deserialize_vec_bytes_from_vec_hex, serialize_vec_bytes_to_hex, Signature,
+    };
+    use crate::subnet_id::SubnetID;
+    use crate::HumanReadable;
+    use fvm_shared::address::Address;
+    use serde::{Deserialize, Serialize};
+    use serde_with::serde_as;
+    use std::str::FromStr;
+
+    #[test]
+    fn test_serialization_vec_vec_u8() {
+        #[serde_as]
+        #[derive(Debug, Serialize, Deserialize, PartialEq)]
+        struct T {
+            #[serde(deserialize_with = "deserialize_vec_bytes_from_vec_hex")]
+            #[serde(serialize_with = "serialize_vec_bytes_to_hex")]
+            d: Vec<Signature>,
+            #[serde_as(as = "HumanReadable")]
+            subnet_id: SubnetID,
+            #[serde_as(as = "HumanReadable")]
+            ipc_address: IPCAddress,
+        }
+
+        let subnet_id =
+            SubnetID::from_str("/r31415926/f2xwzbdu7z5sam6hc57xxwkctciuaz7oe5omipwbq").unwrap();
+        let ipc_address = IPCAddress::new(&subnet_id, &Address::new_id(101)).unwrap();
+
+        let t = T {
+            d: vec![vec![1; 30], vec![2; 30]],
+            subnet_id,
+            ipc_address,
+        };
+        let s = serde_json::to_string(&t).unwrap();
+        assert_eq!(
+            s,
+            r#"{"d":["010101010101010101010101010101010101010101010101010101010101","020202020202020202020202020202020202020202020202020202020202"],"subnet_id":"/r31415926/f2xwzbdu7z5sam6hc57xxwkctciuaz7oe5omipwbq","ipc_address":"/r31415926/f2xwzbdu7z5sam6hc57xxwkctciuaz7oe5omipwbq:f0101"}"#
```

Comment:
Curious: what is that `f0101` at the end?

Reply:
That's the ActorID serialized to string.